server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1314,7 +1297,8 @@ private DeleteResult deleteInLucene(Delete delete, DeletionStrategy plan)
                 if (plan.addStaleOpToLucene || plan.currentlyDeleted) {
```

Comment:
do we need UPDATED_BY_SEQNO_NAME to be updated here too?

Reply:
Good question. A regular delete tombstone is soft-deleted before indexing into Lucene, thus "softUpdateDocument" method won't affect on that document. I think we should not add the updated_by_seqno field to the stale deletes to ensure the consistency between deletes tombstones(i.e., all deletes don't have updated_by_seqno value).

However, I found the case which might be an issue if we remove the safe commit.  This happens as follows:

1. Index a doc (version=1) (seq#0)
2. Delete that doc (seq#1)
3. Index that doc (version=2) (seq#2)

Suppose the global checkpoint is 1, and the seq#2 in the danger zone, we trigger force merge. The problem is seq#2 only updates the updated_by_seqno of  seq#0 because the delete is invisible while seq#0 is still visible (until refresh). A force merge will reclaim the delete and retain seq#0 and seq#2. A rollback seq#2 would restore seq#0 which is incorrect. 

We can avoid this issue if a document is invisible immediately after it's soft-deleted.

Reply:
we discussed this yesterday and this has to do with the fact that lucene treats a document that is index as soft deleted (i.e., the tombstone) differently than a doc that is index and then later soft deleted while that doc is in the indexing buffer. As a work around we decide to detect this and force a refresh to maintain semantics. This will allow this work to continue while we work on a solution in lucene.